File: python code/robots.py
from midlevel import *
from toplevel import *
from lowlevel import *
import numpy as np
import time

class Robot:
    def __init__(self, id, x0):
        self.id = id
        self.x0 = x0 
        # self.t_vals = t_vals
        self.midSystem = None
        self.midControl = None
        self.topSystem = None
        self.lowSystem = None
        self.lowControl = None
        self.mid_layer_time = None
        self.low_layer_time = None
        # self.mid_layer_loop_time = None
        # self.low_layer_loop_time = None

class Robots:
    def __init__(self, idList, x0List, t_vals, cellDict):
        self.idList = idList
        self.numRobots = len(idList)
        self.x0List = x0List
        self.t_vals = t_vals
        self.robotList = []
        self.add_robots()
        self.cellList = cellDict

    def add_robots(self):
        for i in range(self.numRobots):
            id = self.idList[i]
            x0 = self.x0List[i, :]
            self.robotList.append(Robot(id, x0))
        
    def initialise_mid_level_system(self, dim):
        T = self.t_vals.T_mid
        for robot in self.robotList:
            robot.midSystem = MidLevelSystem(robot.id, robot.x0, T, dim)

    def initialise_mid_level_control(self, weight, velocityConstraints, inputConstraints):
        for robot in self.robotList:
            robot.midControl = MidLevelControl(weight, velocityConstraints, inputConstraints, robot.midSystem)

    def initialise_low_level_system(self):
        t_vals = self.t_vals
        for robot in self.robotList:
            x = robot.x0[0]
            y = robot.x0[1]
            # v = np.sqrt(robot.x0[2]**2 + robot.x0[3]**2)
            # Change later if according to original value
            theta = 0
            robot.lowSystem = LowLevelSystem(x, y, theta, t_vals)

    def initialise_low_level_control(self, Kp, Kd, sigma):
        for robot in self.robotList:
            robot.lowControl = LowLevelControl(Kp, Kd, sigma)

    def find_high_level_input(self, sequencePair):
        currentCell = self.cellList[sequencePair[0]]
        terminalCell = self.cellList[sequencePair[1]]
        return (currentCell, terminalCell)
    

File: matlab code/Robots.m
classdef Robots
    properties
        idList
        numRobots

        x0List

        robotList
        cellList 

    end

    methods
        function robots = Robots(idList, x0List, cellDict)
            robots.idList = idList;
            robots.numRobots = length(idList);

            robots.x0List = x0List;

            robots = robots.addRobots();
            robots.cellList = cellDict;

        end

        function robots = addRobots(robots)
            for i = 1:robots.numRobots
                id = robots.idList(i);
                x0 = robots.x0List(:, i);
                robots.robotList = [robots.robotList, Robot(id, x0)];
            end
        end

        % function robots = initialiseRobots(robots, sys, dim, T)
        %     for i = 1:robots.numRobots
        %         robots.robotList(i) = robots.robotList(i).initialiseMidLevelSystem(sys, dim, T);
        %     end
        % end

        function robots = initialiseMidLevelSystem(robots, sys, dim)
            for i = 1:robots.numRobots
                robots.robotList(i).midSystem = MidLevelSystem(robots.robotList(i).id, robots.robotList(i).x0, sys, dim);
            end
        end

        function robots = initialiseMidLevelControl(robots, weight, velocityConstraints, inputConstraints)
            for i = 1:robots.numRobots
                robots.robotList(i).midControl = MidLevelControl(weight, velocityConstraints, inputConstraints, robots.robotList(i).midSystem);
            end
        end

        function highLevelInput = findHighLevelInput(robots, sequencePair)
            currentCell = lookup(robots.cellList, sequencePair(1));
            terminalCell = lookup(robots.cellList, sequencePair(2));
            highLevelInput = {currentCell, terminalCell};
        
        end

    end
end
        
% function midControl = runMPC2(midControl, midSystem)
%     dim = midSystem.dim;

%     % weight = midSystem.weight; 

%     midControl.uc = zeros(dim.nu * dim.N, 1);
%     % midControl.yc = zeros(dim.nx * (dim.N+1), 1);

%     x0 = midSystem.xc;
%     xf = midSystem.xf;

%     u = sdpvar(dim.nu * dim.N, 1);

%     modifiedStateConstraints.A = repelem([midControl.positionConstraints.A; midControl.velocityConstraints.A], dim.N - midControl.counter);
%     modifiedStateConstraints.b = repelem([midControl.positionConstraints.b; midControl.velocityConstraints.b], dim.N - midControl.counter);
%     modifiedTerminalConstraints.A = repelem([midControl.terminalConstraints.A], midControl.counter + 1);
%     modifiedTerminalConstraints.b = repelem([midControl.terminalConstraints.b], midControl.counter + 1);
%     finalStateConstraints.A = [modifiedStateConstraints.A; modifiedTerminalConstraints.A];
%     finalStateConstraints.b = [modifiedStateConstraints.b; modifiedTerminalConstraints.b];
%     finalStateEquation = midSystem.predmod.T * x0 + midSystem.predmod.S * u;

%     constraints = [repelem(midControl.inputConstraints.A, dim.N) * u <= repelem(midControl.inputConstraints.b), ...
%                    finalStateConstraints.A * finalStateEquation <= finalStateConstraints.b];

%     objective = 0.5 * u'*midControl.costmat.H*u + (midControl.costmat.h*(x0-xf))'*u;
%     optimalControl = optimize(constraints, objective);

%     if optimalControl == 0
%         midControl.uc = value(u);
%     else 
%         disp("Optimization problem failed to solve");
%         fprintf("Robot:%s, counter:%d\n", midSystem.id, midControl.counter);
%     end

%     clear u


% end

        

